File: plugin-loader/src/main/java/com/tencent/cubershi/plugin_loader/delegates/DefaultHostActivityDelegate.java
package com.tencent.cubershi.plugin_loader.delegates;

import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.content.Intent;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.util.AttributeSet;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;

import com.tencent.cubershi.mock_interface.MockActivity;
import com.tencent.cubershi.plugin_loader.test.FakeRunningPlugin;
import com.tencent.hydevteam.pluginframework.plugincontainer.HostActivityDelegate;
import com.tencent.hydevteam.pluginframework.plugincontainer.HostActivityDelegator;

import dalvik.system.DexClassLoader;

public class DefaultHostActivityDelegate implements HostActivityDelegate {
    @SuppressWarnings("NullableProblems")
    @NonNull//HostActivityDelegate被传递给HostActivity后应立刻被调用setDelegator()方法传入HostActivityDelegator.
    private HostActivityDelegator mHostActivityDelegator;
    final private DexClassLoader mPluginClassLoader;
    final private Resources mPluginResources;

    public DefaultHostActivityDelegate(DexClassLoader pluginClassLoader, Resources pluginResources) {
        mPluginClassLoader = pluginClassLoader;
        mPluginResources = pluginResources;
    }

    @Override
    public void setDelegator(HostActivityDelegator hostActivityDelegator) {
        mHostActivityDelegator = hostActivityDelegator;
    }

    @Override
    public void onCreate(Bundle bundle) {
        mHostActivityDelegator.superOnCreate(bundle);
        String pluginLauncherActivityName = mHostActivityDelegator.getIntent().getStringExtra(FakeRunningPlugin.ARG);
        try {
            Class<?> aClass = mPluginClassLoader.loadClass(pluginLauncherActivityName);
            MockActivity mockActivity = MockActivity.class.cast(aClass.newInstance());
            mockActivity.setContainerActivity(mHostActivityDelegator);
            mockActivity.setPluginResources(mPluginResources);
            mockActivity.performOnCreate(bundle);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void onResume() {
        mHostActivityDelegator.superOnResume();
    }

    @Override
    public void onNewIntent(Intent intent) {

    }

    @Override
    public void onSaveInstanceState(Bundle bundle) {

    }

    @Override
    public void onPause() {
        mHostActivityDelegator.superOnPause();
    }

    @Override
    public void onStop() {
        mHostActivityDelegator.superOnStop();
    }

    @Override
    public void onDestroy() {
        mHostActivityDelegator.superOnDestroy();
    }

    @Override
    public void onConfigurationChanged(Configuration configuration) {
        mHostActivityDelegator.superOnConfigurationChanged(configuration);
    }

    @Override
    public boolean dispatchKeyEvent(KeyEvent keyEvent) {
        return mHostActivityDelegator.superDispatchKeyEvent(keyEvent);
    }

    @Override
    public void finish() {
        mHostActivityDelegator.superFinish();
    }

    @Override
    public void onActivityResult(int i, int i1, Intent intent) {
        mHostActivityDelegator.superOnActivityResult(i, i1, intent);
    }

    @Override
    public void onChildTitleChanged(Activity activity, CharSequence charSequence) {

    }

    @Override
    public void onRestoreInstanceState(Bundle bundle) {
        mHostActivityDelegator.superOnRestoreInstanceState(bundle);
    }

    @Override
    public void onPostCreate(Bundle bundle) {
        mHostActivityDelegator.superOnPostCreate(bundle);
    }

    @Override
    public void onRestart() {
        mHostActivityDelegator.superOnRestart();
    }

    @Override
    public void onUserLeaveHint() {
        mHostActivityDelegator.superOnUserLeaveHint();
    }

    @Override
    public boolean onCreateThumbnail(Bitmap bitmap, Canvas canvas) {
        return mHostActivityDelegator.superOnCreateThumbnail(bitmap, canvas);
    }

    @Override
    public CharSequence onCreateDescription() {
        return mHostActivityDelegator.superOnCreateDescription();
    }

    @Override
    public Object onRetainNonConfigurationInstance() {
        return mHostActivityDelegator.superOnRetainNonConfigurationInstance();
    }

    @Override
    public void onLowMemory() {
        mHostActivityDelegator.superOnLowMemory();
    }

    @Override
    public boolean onTrackballEvent(MotionEvent motionEvent) {
        return mHostActivityDelegator.superOnTrackballEvent(motionEvent);
    }

    @Override
    public void onUserInteraction() {
        mHostActivityDelegator.superOnUserInteraction();
    }

    @Override
    public void onWindowAttributesChanged(WindowManager.LayoutParams layoutParams) {
        mHostActivityDelegator.superOnWindowAttributesChanged(layoutParams);
    }

    @Override
    public void onContentChanged() {
        mHostActivityDelegator.superOnContentChanged();
    }

    @Override
    public void onWindowFocusChanged(boolean b) {
        mHostActivityDelegator.superOnWindowFocusChanged(b);
    }

    @Override
    public View onCreatePanelView(int i) {
        return mHostActivityDelegator.superOnCreatePanelView(i);
    }

    @Override
    public boolean onCreatePanelMenu(int i, Menu menu) {
        return mHostActivityDelegator.superOnCreatePanelMenu(i, menu);
    }

    @Override
    public boolean onPreparePanel(int i, View view, Menu menu) {
        return mHostActivityDelegator.superOnPreparePanel(i, view, menu);
    }

    @Override
    public void onPanelClosed(int i, Menu menu) {
        mHostActivityDelegator.superOnPanelClosed(i, menu);
    }

    @Override
    public Dialog onCreateDialog(int i) {
        return mHostActivityDelegator.superOnCreateDialog(i);
    }

    @Override
    public void onPrepareDialog(int i, Dialog dialog) {
        mHostActivityDelegator.superOnPrepareDialog(i, dialog);
    }

    @Override
    public void onApplyThemeResource(Resources.Theme theme, int i, boolean b) {
        mHostActivityDelegator.superOnApplyThemeResource(theme, i, b);
    }

    @Override
    public View onCreateView(String s, Context context, AttributeSet attributeSet) {
        return mHostActivityDelegator.superOnCreateView(s, context, attributeSet);
    }

    @Override
    public void startActivityFromChild(Activity activity, Intent intent, int i) {
        mHostActivityDelegator.superStartActivityFromChild(activity, intent, i);
    }

    @Override
    public ClassLoader getClassLoader() {
        return mHostActivityDelegator.superGetClassLoader();
    }
}

File: plugin-loader/src/main/kotlin/com/tencent/cubershi/plugin_loader/delegates/DefaultHostActivityDelegate.kt
package com.tencent.cubershi.plugin_loader.delegates

import android.app.Activity
import android.app.Dialog
import android.content.Context
import android.content.Context.LAYOUT_INFLATER_SERVICE
import android.content.Intent
import android.content.res.Configuration
import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.Canvas
import android.os.Bundle
import android.util.AttributeSet
import android.view.*
import com.tencent.cubershi.mock_interface.MockActivity
import com.tencent.cubershi.mock_interface.MockApplication
import com.tencent.cubershi.plugin_loader.FixedContextLayoutInflater
import com.tencent.cubershi.plugin_loader.infos.PluginActivityInfo
import com.tencent.cubershi.plugin_loader.managers.PluginActivitiesManager
import com.tencent.cubershi.plugin_loader.managers.PluginActivitiesManager.Companion.PLUGIN_ACTIVITY_CLASS_NAME_KEY
import com.tencent.hydevteam.pluginframework.plugincontainer.HostActivityDelegate
import com.tencent.hydevteam.pluginframework.plugincontainer.HostActivityDelegator
import dalvik.system.DexClassLoader

class DefaultHostActivityDelegate(
        private val mPluginApplication: MockApplication,
        private val mPluginClassLoader: DexClassLoader,
        private val mPluginResources: Resources,
        private val mPluginActivitiesManager: PluginActivitiesManager
) : HostActivityDelegate {
    private lateinit var mHostActivityDelegator: HostActivityDelegator
    private lateinit var mMockActivity: MockActivity

    override fun setDelegator(hostActivityDelegator: HostActivityDelegator) {
        mHostActivityDelegator = hostActivityDelegator
    }

    override fun onCreate(bundle: Bundle?) {
        mHostActivityDelegator.superOnCreate(bundle)

        mHostActivityDelegator.intent.setExtrasClassLoader(mPluginClassLoader)

        val bundleForPluginLoader = mHostActivityDelegator.intent.getBundleExtra(PluginActivitiesManager.PLUGIN_LOADER_BUNDLE_KEY)
        mHostActivityDelegator.intent.removeExtra(PluginActivitiesManager.PLUGIN_LOADER_BUNDLE_KEY)
        bundleForPluginLoader.classLoader = this.javaClass.classLoader

        val pluginActivityClassName = bundleForPluginLoader.getString(PLUGIN_ACTIVITY_CLASS_NAME_KEY)
        val pluginActivityInfo: PluginActivityInfo = bundleForPluginLoader.getParcelable(PluginActivitiesManager.PLUGIN_ACTIVITY_INFO_KEY)

        mHostActivityDelegator.setTheme(pluginActivityInfo.themeResource)
        try {
            val aClass = mPluginClassLoader.loadClass(pluginActivityClassName)
            val mockActivity = MockActivity::class.java.cast(aClass.newInstance())
            mockActivity.setContainerActivity(mHostActivityDelegator)
            mockActivity.setPluginResources(mPluginResources)
            mockActivity.setHostContextAsBase(mHostActivityDelegator.hostActivity as Context)
            mockActivity.setPluginClassLoader(mPluginClassLoader)
            mockActivity.setPluginActivityLauncher(mPluginActivitiesManager)
            mockActivity.setPluginApplication(mPluginApplication)
//            mockActivity.setTheme(android.R.style.Theme_DeviceDefault_Light_NoActionBar_Fullscreen)
            mMockActivity = mockActivity
            mockActivity.performOnCreate(bundle)
        } catch (e: Exception) {
            throw RuntimeException(e)
        }

    }

    override fun onResume() {
        mHostActivityDelegator.superOnResume()
    }

    override fun onNewIntent(intent: Intent) {

    }

    override fun onSaveInstanceState(bundle: Bundle) {

    }

    override fun onPause() {
        mHostActivityDelegator.superOnPause()
    }

    override fun onStop() {
        mHostActivityDelegator.superOnStop()
    }

    override fun onDestroy() {
        mHostActivityDelegator.superOnDestroy()
    }

    override fun onConfigurationChanged(configuration: Configuration) {
        mHostActivityDelegator.superOnConfigurationChanged(configuration)
    }

    override fun dispatchKeyEvent(keyEvent: KeyEvent): Boolean {
        return mHostActivityDelegator.superDispatchKeyEvent(keyEvent)
    }

    override fun finish() {
        mHostActivityDelegator.superFinish()
    }

    override fun onActivityResult(i: Int, i1: Int, intent: Intent) {
        mHostActivityDelegator.superOnActivityResult(i, i1, intent)
    }

    override fun onChildTitleChanged(activity: Activity, charSequence: CharSequence) {

    }

    override fun onRestoreInstanceState(bundle: Bundle?) {
        mHostActivityDelegator.superOnRestoreInstanceState(bundle)
    }

    override fun onPostCreate(bundle: Bundle?) {
        mHostActivityDelegator.superOnPostCreate(bundle)
    }

    override fun onRestart() {
        mHostActivityDelegator.superOnRestart()
    }

    override fun onUserLeaveHint() {
        mHostActivityDelegator.superOnUserLeaveHint()
    }

    override fun onCreateThumbnail(bitmap: Bitmap, canvas: Canvas): Boolean {
        return mHostActivityDelegator.superOnCreateThumbnail(bitmap, canvas)
    }

    override fun onCreateDescription(): CharSequence? {
        return mHostActivityDelegator.superOnCreateDescription()
    }

    override fun onRetainNonConfigurationInstance(): Any {
        return mHostActivityDelegator.superOnRetainNonConfigurationInstance()
    }

    override fun onLowMemory() {
        mHostActivityDelegator.superOnLowMemory()
    }

    override fun onTrackballEvent(motionEvent: MotionEvent): Boolean {
        return mHostActivityDelegator.superOnTrackballEvent(motionEvent)
    }

    override fun onUserInteraction() {
        mHostActivityDelegator.superOnUserInteraction()
    }

    override fun onWindowAttributesChanged(layoutParams: WindowManager.LayoutParams) {
        mHostActivityDelegator.superOnWindowAttributesChanged(layoutParams)
    }

    override fun onContentChanged() {
        mHostActivityDelegator.superOnContentChanged()
    }

    override fun onWindowFocusChanged(b: Boolean) {
        mHostActivityDelegator.superOnWindowFocusChanged(b)
    }

    override fun onCreatePanelView(i: Int): View? {
        return mHostActivityDelegator.superOnCreatePanelView(i)
    }

    override fun onCreatePanelMenu(i: Int, menu: Menu): Boolean {
        return mHostActivityDelegator.superOnCreatePanelMenu(i, menu)
    }

    override fun onPreparePanel(i: Int, view: View?, menu: Menu): Boolean {
        return mHostActivityDelegator.superOnPreparePanel(i, view, menu)
    }

    override fun onPanelClosed(i: Int, menu: Menu) {
        mHostActivityDelegator.superOnPanelClosed(i, menu)
    }

    override fun onCreateDialog(i: Int): Dialog {
        return mHostActivityDelegator.superOnCreateDialog(i)
    }

    override fun onPrepareDialog(i: Int, dialog: Dialog) {
        mHostActivityDelegator.superOnPrepareDialog(i, dialog)
    }

    override fun onApplyThemeResource(theme: Resources.Theme, i: Int, b: Boolean) {
        mHostActivityDelegator.superOnApplyThemeResource(theme, i, b)
    }

    override fun onCreateView(name: String, context: Context, attrs: AttributeSet): View? {
        return mHostActivityDelegator.superOnCreateView(name, context, attrs)
    }

    override fun startActivityFromChild(activity: Activity, intent: Intent, i: Int) {
        mHostActivityDelegator.superStartActivityFromChild(activity, intent, i)
    }

    override fun getClassLoader(): ClassLoader {
        return mHostActivityDelegator.superGetClassLoader()
    }

    override fun getLayoutInflater(): LayoutInflater {
        val inflater = mHostActivityDelegator.applicationContext.getSystemService(LAYOUT_INFLATER_SERVICE) as LayoutInflater
        inflater.cloneInContext(mMockActivity)
        return FixedContextLayoutInflater(inflater, mMockActivity)
    }

    override fun getResources(): Resources {
        return mPluginResources
    }
}
